java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/SlidingWindowUtils.java

```diff
@@ -267,11 +278,75 @@ fieldName, getDataType(resolvedSchema, fieldName)))
         final List<DataTypes.Field> fields = new LinkedList<>();
         fields.addAll(keyFields);
         fields.addAll(aggFieldDataTypes);
-        fields.add(
-                DataTypes.FIELD(rowTimeFieldName, getDataType(resolvedSchema, rowTimeFieldName)));
+        fields.add(DataTypes.FIELD(rowTimeFieldName, dataTypeMap.get(rowTimeFieldName)));
         return fields;
     }
 
+    private static Time getModdedOffset(long stepSizeMs, long offsetMs) {
+        offsetMs %= stepSizeMs;
+        if (offsetMs < 0) {
+            offsetMs += stepSizeMs;
+        }
+        return Time.milliseconds(offsetMs);
+    }
+
+    private static class RowKeySelector implements KeySelector<Row, Row> {
+        private final String[] keyFieldNames;
+
+        private RowKeySelector(String[] keyFieldNames) {
+            this.keyFieldNames = keyFieldNames;
+        }
+
+        @Override
+        public Row getKey(Row row) {
+            return Row.of(Arrays.stream(keyFieldNames).map(row::getField).toArray());
+        }
+    }
+
+    private static class RowWindowFunction implements WindowFunction<Row, Row, Row, TimeWindow> {
+        private final AggregationFieldsDescriptor fieldsDescriptor;
+        private final String[] keyFieldNames;
+        private final String rowTimeFieldName;
+
+        private RowWindowFunction(
+                AggregationFieldsDescriptor fieldsDescriptor,
+                String[] keyFieldNames,
+                String rowTimeFieldName) {
+            this.fieldsDescriptor = fieldsDescriptor;
+            this.keyFieldNames = keyFieldNames;
+            this.rowTimeFieldName = rowTimeFieldName;
+        }
+
+        @Override
+        public void apply(
+                Row keys, TimeWindow window, Iterable<Row> iterable, Collector<Row> collector) {
+            Row acc = Row.withNames();
+            for (AggregationFieldsDescriptor.AggregationFieldDescriptor fieldDescriptor :
+                    fieldsDescriptor.getAggFieldDescriptors()) {
+                acc.setField(
+                        fieldDescriptor.outFieldName, fieldDescriptor.aggFunc.createAccumulator());
```

Comment:
It seems weird to output accumulator instance instead of accumulator result. And it might add unexpected serialization overhead since we probably will not add custom serializer for accumulators.

Also note that the existing `RawDataAccumulator` keeps the entire history of records instead of merging them into one record before this accumulator instance is outputted here. This would effectively disable the pre-aggregation optimization used before this PR and considerably increase the memory usage.

Instead of putting accumulator instances in the output row, would it be more intuitive to output concrete values (e.g. accumulated results) similar to the approach prior to this PR?